### PR TITLE
Related content block for Taxonomy tags on content.

### PR DESCRIPTION
**Pull request by Shinsina**
LIVE
Screen Shot 2021-09-15 at 10 39 05 AM

DEV
Screen Shot 2021-09-15 at 10 38 47 AM

This will "backfill" the related content block with content found with the one of the same tags as the current content page, if the block already has 4 pieces of content (listed as relatedTo on a content item) it will simply render those 4 items.

### CODE REVIEW

**Comment by Shinsina** [comment]
[Add support for a Related Content block using taxonomy terms](https://trello.com/c/j6S69Mqq/161-add-support-for-a-related-content-block-using-taxonomy-terms)

File: packages/shared/components/blocks/related-content.marko [hunk 1]
```diff
@@ -1,19 +1,61 @@
...
+$ const includeTaxonomyIds = input.taxonomyIds;
+
 <marko-web-query|{ nodes }|
   name="related-published-content"
   params={ contentId: input.contentId, limit: 4, withSite: false, queryFragment }
 >
-  <shared-content-list-flow
+  $ const excludeContentIds = nodes.map((node) => node.id)
+  $ excludeContentIds.push(input.contentId);
+  <if(nodes.length < 4 && includeTaxonomyIds.length)>
```
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:19]: I would move the `limit` value to a constant above, then do `nodes.length < limit`. Also, L15 should also use the global limit

File: packages/shared/components/blocks/related-content.marko [hunk 2]
```diff
@@ -1,19 +1,61 @@
...
+    $ allRelatedNodes.push(...nodes)
+    <marko-web-query|{ nodes }|
+      name="all-published-content"
+      params={
+        excludeContentTypes,
+        excludeContentIds,
+        includeTaxonomyIds,
+        limit: 4-nodes.length,
```
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:27]: Change to `limit: limit - nodes.length`

File: packages/shared/components/blocks/related-content.marko [hunk 3]
```diff
@@ -1,19 +1,61 @@
...
+        withSite: false,
+        queryFragment
+      }
+      >
```
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:31]: Fix indent

File: packages/shared/components/blocks/related-content.marko [hunk 4]
```diff
@@ -1,19 +1,61 @@
 import queryFragment from "../../graphql/fragments/related-content-block";
 
+$ const excludeContentTypes = [
+  "Contact",
+  "Company",
+  "Document",
+  "TextAd",
+];
+
+$ const allRelatedNodes = [];
```
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:10]: This constant is only used on L19... as such, move it's declaration to L20

File: packages/shared/components/blocks/related-content.marko [hunk 5]
```diff
@@ -1,19 +1,61 @@
...
+$ const includeTaxonomyIds = input.taxonomyIds;
```
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:11]: Destructure this (and `contentId`)...

```marko
$ const { taxonomyIds, contentId } = input;
```

Then change L15 to `params={ contentId, limit ... }`
Change L18 to `$ excludeContentIds.push(contentId);`
Change L19 to `&& taxonomyIds.length`
And L26 to `includeTaxonomyIds: taxonomyIds`

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: packages/shared/components/blocks/related-content.marko (hunk 1 above)
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:19]: Also, there is no default value for `taxonomyIds` so the `.length` call will error if a non-array is passed. As such, you should use the `getAsArray` utility function

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: packages/shared/components/blocks/related-content.marko [hunk 6]
```diff
@@ -1,19 +1,64 @@
...
+$ const limit = 4;
+
 <marko-web-query|{ nodes }|
   name="related-published-content"
-  params={ contentId: input.contentId, limit: 4, withSite: false, queryFragment }
+  params={ contentId, limit, withSite: false, queryFragment }
 >
-  <shared-content-list-flow
+  $ const allRelatedNodes = [];
+  $ const excludeContentIds = nodes.map((node) => node.id)
```
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:20]: Add semi-colons to the end of L20, L23, and L35

File: packages/shared/components/blocks/related-content.marko [hunk 7]
```diff
@@ -1,19 +1,64 @@
...
+    $ allRelatedNodes.push(...nodes)
+    <marko-web-query|{ nodes }|
+      name="all-published-content"
+      params={
+          excludeContentTypes,
+          excludeContentIds,
+          includeTaxonomyIds: taxonomyIds,
+          limit: limit-nodes.length,
+          withSite: false,
+          queryFragment
```
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:32]: Add trailing comma

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: packages/shared/components/blocks/related-content.marko [hunk 8]
```diff
@@ -1,19 +1,64 @@
...
+  $ excludeContentIds.push(contentId);
+  <if(nodes.length < limit && taxonomyIds.length)>
+    $ allRelatedNodes.push(...nodes)
+    <marko-web-query|{ nodes }|
+      name="all-published-content"
+      params={
+          excludeContentTypes,
+          excludeContentIds,
+          includeTaxonomyIds: taxonomyIds,
+          limit: limit-nodes.length,
```
Comment (zarathustra323) [inline packages/shared/components/blocks/related-content.marko:30]: Include spaces between the minus sign... `limit - nodes.length`

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

**Review by zarathustra323 — APPROVED** [review]

**Pull request merged**